draft-ietf-ppm-dap.md

```diff
@@ -1377,20 +1389,30 @@ This message consists of:
 * `prepare_inits`: the sequence of `PrepareInit` messages constructed in the
   previous step.
 
-Finally, the Leader sends a PUT request to
-`{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}`. The payload
-is set to `AggregationJobInitReq` and the media type is set to
-"application/dap-aggregation-job-init-req".
+Finally, the Leader sends an HTTP PUT request to
+`{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` with a media
+type of "application/dap-aggregation-job-init-req" and a body containing the
+`AggregationJobInitReq`.
 
 The Leader MUST authenticate its requests to the Helper using a scheme that
 meets the requirements in {{request-authentication}}.
 
-The Helper's response will be an `AggregationJobResp` message (see
-{{aggregation-helper-init}}. The response's `prepare_resps` must include exactly
-the same report IDs in the same order as the Leader's `AggregationJobInitReq`.
-Otherwise, the Leader MUST abort the aggregation job.
-
-[[OPEN ISSUE: consider relaxing this ordering constraint. See issue#217.]]
+The Helper responds with HTTP status 201 Created. If the response has a media
+type of "application/dap-aggregation-job-resp" with a body containing an
+`AggregationJobResp`, the Leader proceeds onward (see
+{{aggregation-helper-init}}). Otherwise, the Leader polls the aggregation job
+by sending HTTP GET requests to
+`{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}?step=0`. If the
+job is not ready, the Helper responds to the GET request with HTTP status 202
+Accepted with no media type and body. The Helper's response SHOULD include a
```

Comment:
202 is not the right status code to return from polling.  If the goal is to have a resource that indicates the status of the job, it should always return 200 with a body that indicates what the status is.

Reply:
Got it, thanks for catching this.

When the step is ready, the Helper returns 200 OK with an `AggregationJobResp`, so we would need to disambiguate this "ready" response from a "not ready" response. Could we use the content types for this? Something like:

* "ready": content = [application/dap-aggregation-job-resp](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-8.1.4); body = [`AggregationJobResp`](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-4.7.1.2-19)
* "not ready": content = application/dap-aggregation-job-pending (or something); body = empty

Or perhaps it would be more correct to change `AggregationJobResp` to convey whether the step is ready?

Reply:
> 202 is not the right status code to return from polling. If the goal is to have a resource that indicates the status of the job, it should always return 200 with a body that indicates what the status is.

That tracks with prior art https://www.rfc-editor.org/rfc/rfc8555#section-7.5.1. (How embarrassing! I should have consulted ACME far earlier :wink:).

I have added a `status` field to the `AggregationJobResp`. I think this brings nice text simplifications, because we express the mutual exclusion between a processing job and the presence of prep_resps through the struct definition.